### PR TITLE
2.0.0 Changes For Bookshelf History

### PR DESCRIPTION
**Pull request by HoyaBoya**
# Description

This updates Bookshelf History to 1.1.0 with the following changes:

* Ran `standardize --fix` on the project.
* Updated bookshelf pin to 1.1.0
* We now keep track of model `changed` by running a JSON diff of the previous attributes and current attributes.
* Selective hooks depending on if it is a `updated` or `updating`.
* Updated documentation
* UPDATE: Added support for function callbacks for metdata.

/cc @bsiddiqui

### CODE REVIEW

**Comment by HoyaBoya** [comment]
A few updates

* Remove xtend Module and used `Object.assign`.
* Cleaned up how we add the `updating` hook.
* Researched model.options.previousAttributes() and it didn't work 😢 
* Warn when not in a transaction
* Error when not using a created or updated hook.

File: index.js [hunk 1]
```diff
@@ -156,27 +177,37 @@ module.exports = (bookshelf, options) => {
...
-      let fields = model.historyOptions.fields
-      let execute = transacting => {
-        let forge = {}
+      const fields = model.historyOptions.fields
+      const execute = transacting => {
+        const forge = {}
         forge[fields.resource_id] = model.id
         forge[fields.resource_type] = model.tableName
 
+        if (!transacting) {
```
Comment (HoyaBoya) [inline index.js]: @bsiddiqui is this the right place to put the warning?

File: index.js [hunk 2]
```diff
@@ -9,20 +9,21 @@ let extend = require('xtend')
...
  * @return {Object} A new base model that can be extended
  */
-module.exports = (bookshelf, options) => {
-  let base = bookshelf.Model
-  let defaults = extend({
-    fields: {
+module.exports = (bookshelf, options = {}) => {
+  const base = bookshelf.Model
+  const defaults = {
+    fields: Object.assign(options.fields || {}, {
```
Comment (HoyaBoya) [inline index.js]: Is this correct?

File: index.js [hunk 3]
```diff
@@ -34,21 +35,41 @@ module.exports = (bookshelf, options) => {
 
       // Do nothing if the model doesn't have history enabled
       if (this.history) {
-        this.historyOptions = extend(defaults, this.history)
+        this.historyOptions = Object.assign(defaults, this.history)
```
Comment (HoyaBoya) [inline index.js]: Is this correct?

**Comment by HoyaBoya** [comment]
@bsiddiqui returned `xtend`.

File: index.js [hunk 4]
```diff
@@ -39,16 +44,37 @@ module.exports = (bookshelf, options) => {
         return
       }
 
+      const hookMap = {
+        created: 'creating',
```
Comment (concertcoder) [inline index.js]: Maybe we want to keep track of destroying as well?

File: index.js [hunk 5]
```diff
@@ -39,16 +44,37 @@ module.exports = (bookshelf, options) => {
...
       if (this.historyOptions.autoHistory) {
         this.historyOptions.autoHistory.forEach(hook => {
-          this.on(hook, (model, attrs, options) => {
-            if (options.history === false) {
-              return
-            } else {
-              return model.constructor.history(model, Boolean(options.patch), hook, options.transacting)
-            }
-          })
+          if (Object.keys(hookMap).includes(hook)) {
```
Comment (concertcoder) [inline index.js]: If else is redundant, as we check for this right above

File: index.js [hunk 6]
```diff
@@ -152,31 +178,49 @@ module.exports = (bookshelf, options) => {
...
+            const data = {}
+            data[fields.sequence] = sequence
+            data[fields.resource_id] = model.id
+            data[fields.resource_type] = model.tableName
+            data[fields.data] = JSON.stringify(model.attributes)
+            data[fields.patch] = Boolean(patch)
+            data[fields.operation] = operation
+
+            if (historyOptions.authorCallback && typeof historyOptions.authorCallback === 'function') {
+              const metadata = historyOptions.authorCallback(model)
```
Comment (concertcoder) [inline index.js]: Not sure I love the naming of the function "authorCallback".  usually you start with a verb so I'm thinking of authoring.  Maybe something like "getAuthorMetadata" or something

File: test/migrations/20200228112321_create_history.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      .notNullable()
+      .defaultTo(knex.raw('uuid_generate_v4()'))
+    t.integer('sequence').notNullable()
+    t.string('operation').notNullable()
+    t.boolean('patch').notNullable()
+    t.string('resource_type').notNullable()
+    t.uuid('resource_id').notNullable()
+    t.string('author_type')
+    t.uuid('author_id')
+    t.jsonb('changed')
```
Comment (concertcoder) [inline test/migrations/20200228112321_create_history.js]: Maybe diff instead of changed

**Comment by HoyaBoya** [comment]
@concertcoder all feedback addressed.

File: index.js [hunk 8]
```diff
@@ -189,7 +189,9 @@ module.exports = (bookshelf, options = {}) => {
           console.warn('No transaction detected at time of History write. In the future, Bookshelf-History will enforce a transaction.')
         }
 
-        return model.historyOptions.model.forge(forge)
+        return model.historyOptions.model
+          .forge(forge)
+          .query(qb => qb.orderBy('sequence', 'desc'))
```
Comment (HoyaBoya) [inline index.js]: @bsiddiqui this touch commit contains the fix we worked on IRL.

File: index.js [hunk 9]
```diff
@@ -39,18 +44,36 @@ module.exports = (bookshelf, options) => {
         return
       }
 
+      const hookMap = {
+        created: 'creating',
+        updated: 'updating',
+        destroyed: 'destroying'
```
Comment (bsiddiqui) [inline index.js]: this is a bit confusing - why are you mapping `ed` =>`ing`?

File: index.js (hunk 1 above)
Comment (bsiddiqui) [reply index.js]: yeah this works

File: index.js [hunk 10]
```diff
@@ -152,31 +175,51 @@ module.exports = (bookshelf, options) => {
...
+            const data = {}
+            data[fields.sequence] = sequence
+            data[fields.resource_id] = model.id
+            data[fields.resource_type] = model.tableName
+            data[fields.data] = JSON.stringify(model.attributes)
+            data[fields.patch] = Boolean(patch)
+            data[fields.operation] = operation
+
+            if (historyOptions.getAuthorMetadata && typeof historyOptions.getAuthorMetadata === 'function') {
+              const metadata = historyOptions.getAuthorMetadata(model)
```
Comment (bsiddiqui) [inline index.js]:
```js
const metadata = {
  author_id: 1,
  author_type: 'users'
}

if (metadata) {
  data = { ...data, ...metadata } 
}

File: index.js [hunk 11]
```diff
@@ -152,31 +175,51 @@ module.exports = (bookshelf, options) => {
...
+          console.warn('No transaction detected at time of History write. In the future, Bookshelf-History will enforce a transaction.')
+        }
+
+        return model.historyOptions.model
+          .forge(forge)
+          .query(qb => qb.orderBy('sequence', 'desc'))
+          .fetch({ transacting, require: false })
+          .then(row => row ? Number(row.get(fields.sequence)) + 1 : 1)
+          .then(sequence => {
+            const data = {}
```
Comment (bsiddiqui) [inline index.js]:
```js
const data = {
  // default values
  [fields.sequence]: sequence
  ...options.getAdditionalValues()
}

await model.save(data)

File: test/index.js [hunk 12]
```diff
@@ -1,8 +1,10 @@
 'use strict'
```
Comment (HoyaBoya) [inline test/index.js]: @bsiddiqui, I rewrote all the specs for the latest Lab and Code, using `await / async`. 

```
> bookshelf-history@1.0.0 test /Users/christophersun/Projects/homevest/bookshelf-history
> standard && lab --verbose --colors --assert @hapi/code --coverage

bookshelf-history
  ✔ 1) it runs with default arguments (22 ms and 5 assertions)
  ✔ 2) it detects patch operation (10 ms and 5 assertions)
  ✔ 3) it does not create tuples when history is disabled (2 ms and 2 assertions)
  ✔ 4) it disables backup when options.history is false (6 ms and 3 assertions)
backup()
  ✔ 5) it saves a manual history (6 ms and 2 assertions)
  ✔ 6) it accepts running transactions (5 ms and 2 assertions)
revert()
  ✔ 7) it reverts (15 ms and 5 assertions)
  ✔ 8) it accepts running transactions (11 ms and 5 assertions)
  ✔ 9) it uses latest sequence when none is provided (11 ms and 2 assertions)
  ✔ 10) it works with json datatypes on postgres (18 ms and 5 assertions)
history
  ✔ 11) works (8 ms and 4 assertions)
```

File: test/index.js [hunk 13]
```diff
@@ -1,8 +1,10 @@
...
 
-const Lab = require('lab')
+require('dotenv').config()
```
Comment (HoyaBoya) [inline test/index.js]: I bring in `dotenv` to read the DATABASE_URL if not passed in. It is a dev-only dependency.

File: index.js [hunk 14]
```diff
@@ -50,6 +50,10 @@ module.exports = (bookshelf, options = {}) => {
         destroyed: 'destroying'
       }
 
+      if (!Array.isArray(this.historyOptions.autoHistory)) {
```
Comment (HoyaBoya) [inline index.js]: This fixes a bug where `false` and `null` where being passed in for `autoHistory`.

File: index.js [hunk 15]
```diff
@@ -190,7 +194,7 @@ module.exports = (bookshelf, options = {}) => {
 
         return model.historyOptions.model
           .forge(forge)
-          .query(qb => qb.orderBy('sequence', 'desc'))
+          .query(qb => qb.orderBy(fields.sequence, 'desc'))
```
Comment (HoyaBoya) [inline index.js]: This fixes a bug where we need the configured `sequence`, not the hard coded value.

**Pull request merged**